docs/customization.md

```diff
@@ -0,0 +1,167 @@
+parameterless constructor.
+
+When writing integration tests, it's often important to be able to
+swap out your "production" Functions Startup classes with "test"
+versions, for example to use fakes or mocks for services configured
+via dependency injection. The `FunctionTestServerBuilder` makes this
+easy by allowing Functions Startup instances to be specified while
+building the test server. See [the testing
+documentation](testing.md) for more details.
+
+## Customizing Dependency Injection using `ConfigureServices`
+
+The Functions Framework Invoker uses [ASP.NET Core dependency
+injection](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/dependency-injection)
+when constructing instances of the target function type to handle
+calls.
+
+See [SimpleDependencyInjection.Function](../examples/Google.Cloud.Functions.Examples.SimpleDependencyInjection/Function.cs)
+for an example of this, where the function constructor has an
+`ILogger<T>` parameter so that a logger can be injected.
+
+In more complex scenarios, users may wish to provide additional
+dependencies themselves through Functions Startup classes, by
+overriding the `ConfigureServices` method:
+
+```csharp
+public virtual void ConfigureServices(WebHostBuilderContext context, IServiceCollection services)
+```
+
+The dependencies are then applied by ASP.NET Core dependency
+injection in the normal way. The 
+
+See [AdvancedDependencyInjection.Function](../examples/Google.Cloud.Functions.Examples.AdvancedDependencyInjection/Function.cs)
+for an example of using scoped and singleton services within a single function. Each function invocation uses
+a new instance of its scoped dependency, but uses the same instance of the singleton dependency.
+
+## Customizing the request pipeline using `Configure`
+
+The `Configure` method in the Functions Startup class is provided
+with the web host builder context and the application builder, as it
+would be if you called `IWebHostBuilder.Configure` with a
+configuration delegate:
+
+```csharp
+public virtual void Configure(WebHostBuilderContext context, IApplicationBuilder app)
+```
+
+The code in the Functions Startup is run during application
+configuration, before the Functions Framework configures its
+own request handling. This allows you to add additional middleware.
+Note that while nothing prevents you from installing arbitrary
+middleware that terminates the pipeline early, the expectation is
+that successful requests will always end up reaching the function.
+(For example, installing middleware that terminates the request with
+an error for certain conditions is fine; installing middleware to
+respond with files from the file system is less conventional.)
+
+See [Middleware.Function](../examples/Google.Cloud.Functions.Examples.Middleware/Function.cs)
+for an example of a small piece of middleware providing extra per-request logging.
+
+## Customizing application configuration sources using `ConfigureAppConfiguration`
```

Comment:
I suppose you explicitly orderd these sections by importance or provability of use. I think that's fine, just calling out here that maybe a note on execution order is relevant.
For instance, it's relevant that configuration added here is available for everything else, etc.

Reply:
Good plan, done.